### ISSUE TITLE
SSL_write succeeds and does not report cert verification errors with TLS 1.3 (may result in data loss)

### ISSUE THREAD

**Issue by szemere** [openssl/openssl#13998]
Hello,

 I have a one-way protocol where the client only uses the SSL_write method and never calls SSL_read. This setup works really well with TLSv1.2 (and older), however with TLSv1.3 the client is unable to detect error cases.

 By browsing the already opened issues I have found similar, but not identical cases. They include session tickets, mid communication key renewal, partial messages in the buffers, etc. The answers in those cases are suggesting: "Assuming two way protocols, SSL_read will report the problem."

 By inserting read methods into my code[1] - **I can confirm** the errors are reported immediately! However I think there is a chance for data loss even in case of two-way protocols, if the client is the one who sends data first. This is the reason I am opening this issue.

Please help me identify if this is a bug, or TLSv1.3 requires a fundamentally different architecture. In case it's not considered as a bug, I think it would be important to add something to the documentation that explains what to do when implementing one-way protocols/upgrading an existing implementation to TLSv1.3 (OpenSSL 1.1).

# Envrionment

```
$ openssl version -a
OpenSSL 1.1.1  11 Sep 2018
built on: Wed Dec  2 14:54:45 2020 UTC
platform: debian-amd64
options:  bn(64,64) rc4(16x,int) des(int) blowfish(ptr) 
compiler: gcc -fPIC -pthread -m64 -Wa,--noexecstack -Wall -Wa,--noexecstack -g -O2 -fdebug-prefix-map=/build/openssl-WsH1Tl/openssl-1.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DKECCAK1600_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DX25519_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DNDEBUG -Wdate-time -D_FORTIFY_SOURCE=2
OPENSSLDIR: "/usr/lib/ssl"
ENGINESDIR: "/usr/lib/x86_64-linux-gnu/engines-1.1"
Seeding source: os-specific
```

```
$ uname -a
Linux tp 5.4.0-62-generic #70~18.04.1-Ubuntu SMP Tue Jan 12 17:18:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

# Reproduction code

https://github.com/szemere/openssl-missingcert-test/blob/master/ssl_write_test.c

I was able to reproduce the issue with any case where the server provides the "SSL routines:tls_process_client_certificate:certificate verify failed" error message. So far I tested:
- expired certificates
- omitting the client certificate
- providing invalid certificate

https://github.com/szemere/openssl-missingcert-test/blob/master/ssl_server_for_testing.c

First, I tested with `openssl s_server`, which gives the same result. But in order to turn off session tickets[2] I provided a simple "read everything" server so the reproduction is self contained.

# Reproduction steps

1. Generate proper certificates for the server.
2. Start the provided example or the `openssl s_server` (my options: [3])
3. Start the provided client.

Output on the server side reports the issue with the certificate properly and closes the connection:
```
140681764807104:error:1417C086:SSL routines:tls_process_client_certificate:certificate verify failed:../ssl/statem/statem_srvr.c:3668:
```

Output on the client side suggests successful data transer:
```
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: -1 EAGAIN
SSL_write returned: 512 SUCCESS
```

My client exits if SSL_write returns any value bigger than 0.

- Any following SSL_write also succeeds, reporting succesful data transfer. (I tested it with an infinite loop around the `try_sending_512bytes` method.)
- The first SSL_read properly reports the problem, as stated above. [4]
- SSL_write already reported successful data transer, so data loss is already possible.
- The following `SSL_shutdown` method reports 0.

# Notes

[1] Inserting read methods requires a protocol change.

[2] As a workaround to: https://github.com/openssl/openssl/issues/7948

[3] Server command: `openssl s_server -Verify 42 -verify_return_error -tls1_3 -cert /tmp/server.crt -key /tmp/server.key -verifyCApath /tmp/CAdir/`

[4] Of course, in order to call SSL_read before the first SSL_write, it is required to call SSL_connect explicitly. Unfortunately SSL_connect also reports success.

I only tested the setup with a certificate error during the handshake process. At this moment I am unaware of any other connection error which causes the same effect.

**Comment by richsalz** [comment]
I think you're using TLS wrong.  The protocol messages need to flow in two directions.  They always did.

**Comment by mattcaswell** [comment]
A TLSv1.3 handshake looks like this (from RFC 8446):

````
       Client                                           Server

Key  ^ ClientHello
Exch | + key_share*
     | + signature_algorithms*
     | + psk_key_exchange_modes*
     v + pre_shared_key*       -------->
                                                  ServerHello  ^ Key
                                                 + key_share*  | Exch
                                            + pre_shared_key*  v
                                        {EncryptedExtensions}  ^  Server
                                        {CertificateRequest*}  v  Params
                                               {Certificate*}  ^
                                         {CertificateVerify*}  | Auth
                                                   {Finished}  v
                               <--------  [Application Data*]
     ^ {Certificate*}
Auth | {CertificateVerify*}
     v {Finished}              -------->
       [Application Data]      <------->  [Application Data]
````
Each peer considers the handshake to be complete when it has both sent and successfully received/validated a "Finished" message. You can see from the above diagram that by the time the client sends its "Finished" message it has already received and validated the server one. Therefore the client considers the handshake complete immediately after it has sent its Finished. The server meanwhile is still waiting on the final flight from the client, and therefore the server hasn't finished yet.

The error you are talking about occurs at the point in the protocol after the client believes it is done, but the server is still processing the handshake.

This is just the way TLSv1.3 works. If you have a "one way protocol" then the client reallly *must* not assume everything is successful until a full 2-way shutdown has been performed (meaning that it has both sent *and* received a close_notify alert). From an API perspective this means you must continue calling `SSL_shutdown` until it has returned 1 (or indicates an error).

https://www.openssl.org/docs/man1.1.1/man3/SSL_shutdown.html

Note, that this problem also exists in TLSv1.2 when using session resumption so it is not new to TLSv1.3. The main difference is that it occurs in TLSv1.3 for non-resumed sessions too.

**Comment by szemere** [comment]
Hello @mattcaswell,
 Thanks for the detailed answer.

One of the mentioned "one-way protocols" is [RFC5425](https://tools.ietf.org/html/rfc5425).

If I understand you correctly, due to the "async" nature of the TLS 1.3 handshake, it is now possible to lose all of the
application-level payload in RFC5425, because clients never read data (AFAIK the same is true for FTPS data channels).

The only thing an implementation of this can do is calling SSL_shutdown() until both close_notifys are delivered/received.
Well, SSL_shutdown() is too late for us, as we'd like to detect message loss as soon as possible (which is difficult with a protocol without application-level acks), but as a workaround, can we call SSL_read() to read 0 "application-level bytes" to detect the cert verification failure reliably? (This seems like a very odd workaround. i.e.: SSL_write() states that calling with num=0 is an undefined behaviour.)

(Sorry if it's out-of-scope, I just want to make sure I understood everything well.)

**Comment by kaduk** [comment]
It has always been the case with TLS that to avoid truncation of the data stream (or "data loss" as you put it), either the application layer protocol needs to indicate message sizes and successful completion, or you have to get and process the "close_notify" alerts in both directions.  Since a one-way application protocol inherently can't report back completion in the other direction, that means you have to use the "close_notify" procedure, which in OpenSSL is implemented via the SSL_shutdown() routine.  I'm sorry that this was not clear from the documentation, but it doesn't change the reality that this risk has always been present in your software.

**Comment by MrAnno** [comment]
Hi @kaduk,

Thank you for your response.

I agree, it's a fact that RFC5425 has no per-message delivery guarantees due to the simplicity of the protocol, so truncation is possible without processing both close_notifys (as stated in 4.4.  Closure).

After reading https://github.com/openssl/openssl/issues/13998#issuecomment-769082091, we were just wondering if it was possible to detect the certificate verification failure earlier than the end of the connection without reading actual app-level data.

**Comment by mattcaswell** [comment]
> After reading #13998 (comment), we were just wondering if it was possible to detect the certificate verification failure earlier than the end of the connection, without reading actual app-level data.

You can attempt to call SSL_read() and you can do so with a buffer larger than 0. Since you're not actually expecting a reply the buffer will never be filled in anyway (and presumably it would be an error in your protocol if you did get such a reply). Calling SSL_read() when you're not actually expecting a reply will result in one of two things:

1) Non fatal error (SSL_ERROR_WANT_READ)
2) Some fatal error e.g. SSL_ERROR_SSL/SSL_ERROR_SYSCALL

In the case of (2) you know that the server has failed (possibly because of the cert verify failure but it could be for some other reason). In the case of (1) you don't know whether the server has failed, or is going to fail and just hasn't go around to telling you yet, or whether all is good.

So, calling SSL_read(), *might* tell you about the failure earlier than you would otherwise know via SSL_shutdown, but there is no guarantee and no response cannot be interpreted as "all is good".
